src/ast/json.rs

```diff
@@ -22,6 +22,60 @@ const CHAR_ARRAY: char = 'a';
 const CHAR_OBJECT: char = 'o';
 const CHAR_STRING: char = 's';
 
+/// The sapling representation of the AST for a subset of JSON (where all values are either 'true'
+/// or 'false', and keys only contain ASCII).
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub enum Class {
+    /// True
+    True,
+    /// False
+    False,
+    /// Null
+    Null,
+    /// Str
+    Str,
+    /// Array
+    Array,
+    /// Object
+    Object,
+}
+
+impl<'arena> NodeClass<'arena> for Class {
+    fn to_char(self) -> char {
+        match self {
+            Class::True => 't',
+            Class::False => 'f',
+            Class::Null => 'n',
+            Class::Array => 'a',
+            Class::Object => 'o',
+            Class::Str => 'c',
```

Comment:
I think that these `char` literals should be `CHAR_TRUE`, `CHAR_FALSE`, etc.

At any rate, this whole thing is asking to be turned into a macro, which would fix the issue.  I'd quite like to do that because I'd like to learn more about macros, but if you want to have a go at doing it now then that's fine too.

Reply:
Actually macros is the only chapter I skipped while reading the book.

Reply:
Ah OK that's fair - macros are one of those things that aren't useful 99% of the time, but when you do need them they are incredibly useful.  I'll make a macro for this as a separate PR (unless you want to).